bin/dm_parse_ea.py

```diff
@@ -1,294 +1,504 @@
+        {
+            "onset": log["Time"].loc[rating_mask].values,
+            "participant_value": log.loc[rating_mask]["Code"].values,
+            "event_type": "button_press",
+            "duration": 0,
+        }
+    )
 
-    #gets rating substring from participant numbers
-    df['participant_value'] = df['participant_value'].str.strip().str[-1]
+    # gets rating substring from participant numbers
+    df["participant_value"] = df["participant_value"].str.strip().str[-1]
 
-    return(df)
+    return df
 
-#combines the block rows with the ratings rows and sorts them
-def combine_dfs(blocks,ratings):
-    combo=blocks.append(ratings).sort_values("onset").reset_index(drop=True)
 
-    mask = pd.notnull(combo['trial_type'])
+# combines the block rows with the ratings rows and sorts them
+def combine_dfs(blocks, ratings):
+    combo = blocks.append(ratings).sort_values("onset").reset_index(drop=True)
+    mask = pd.notnull(combo["trial_type"])
+    combo["space_b4_prev"] = combo["onset"].diff(periods=1)
+    combo["first_button_press"] = combo["duration"].shift() > 0
+    combo2 = combo.drop(
+        combo[
+            (combo["space_b4_prev"] < 1000)
+            & (combo["first_button_press"] is True)
+        ].index
+    ).reset_index(drop=True)
 
-    combo['rating_duration']=combo['onset'].shift(-1)-combo['onset'].where(mask==False)
+    mask = pd.notnull(combo2["trial_type"])
 
-    block_start_locs=combo[mask].index.values
+    block_start_locs = combo2[mask].index.values
 
+    last_block = combo2.iloc[block_start_locs[len(block_start_locs) - 1]]
 
-    #TODO: fix this lol
-    #this ends up not assigning a value for the final button press - there must be a more elegant way to do all this
-    for i in range(len(block_start_locs)):
-        if block_start_locs[i] != 0:
-            #maybe i should calculate these vars separately for clarity
-            combo.rating_duration[block_start_locs[i]-1]=combo.end[block_start_locs[i-1]] - combo.onset[block_start_locs[i]-1]
+    end_row = {
+        "onset": last_block.end,
+        "rating_duration": 0,
+        "event_type": "last_row",
+        "duration": 0,
+        "participant_value": last_block.participant_value,
+    }
 
+    combo2 = combo2.append(end_row, ignore_index=True).reset_index(drop=True)
 
-#adds rows that contain the 5 second at the beginning default value
-    for i in block_start_locs:
-            new_row={'onset':combo.onset[i],
-            'rating_duration':combo.onset[i+1] - combo.onset[i],
-            'event_type':'default_rating',
-            'duration':0,
-            'participant_value':5}
-            combo=combo.append(new_row,ignore_index=True)
+    mask = pd.notnull(combo2["trial_type"])
 
-    combo=combo.sort_values("onset").reset_index(drop=True)
+    block_start_locs = combo2[mask].index.values
 
-    return(combo)
+    combo2["rating_duration"] = combo2["onset"].shift(-1) - combo2[
+        "onset"
+    ].where(mask is False)
```

Comment:
```suggestion
    combo2["rating_duration"] = combo2["onset"].shift(-1) - combo2[
        "onset"
    ].where(mask)
```

I get an error w/pandas 1.0.3 since "mask is False" evaluates to a single element

Reply:
also I think `DataFrame.diff()` can perform the above operation in a much cleaner fashion?